### PR TITLE
test: Add -k, --keep option for compiled binaries

### PR DESCRIPTION
**Pull request by ggstone0523**
Add an option to keep compiled binaries 
so that runtest.py do not remove test directory if using -k or --keep option

Fixed: #1469

Signed-off-by: Yusun Choi <tanks2438@outlook.kr>

### CODE REVIEW

**Comment by kangtegong** [comment]
Thanks for your contribution but I think this needs a little bit more amends.

First, you should commit message header from `test:Add -k, --keep option for compiled binaries
` to `tests: Add -k, --keep option for compiled binaries`

Second, please change `Fixed: #[1469]` to `Fixed: #1469` in your commit message.

**Comment by ggstone0523** [comment]
Thanks you for comment! I amend commit message following the comment.

**Review by honggyukim — COMMENTED** [review]
The current change leaves the test directory under `/tmp` so better to move it to the current directory. Could you apply this change?
```diff
diff --git a/tests/runtest.py b/tests/runtest.py
index 43e6711d..ea6faa67 100755
--- a/tests/runtest.py
+++ b/tests/runtest.py
@@ -73,8 +73,6 @@ class TestBase:
         'C++': { 'cc': 'g++', 'flags': 'CXXFLAGS', 'ext': '.cpp' },
     }

-    keep = False
-
     TEST_SUCCESS = 0
     TEST_UNSUPP_LANG = -1
     TEST_BUILD_FAIL = -2
@@ -95,6 +93,8 @@ class TestBase:
     feature = set()

     def __init__(self, name, result, lang='C', cflags='', ldflags='', sort='task', serial=False):
+        self.keep = False
+        self.orig_dir = os.getcwd()
         _tmp = tempfile.mkdtemp(prefix='test_%s_' % name)
         os.chdir(_tmp)
         self.test_dir = _tmp
@@ -628,7 +628,9 @@ class TestBase:
         return ret, ''

     def __del__(self):
-        if not self.keep:
+        if self.keep:
+            sp.call(['mv', self.test_dir, self.orig_dir])
+        else:
             sp.call(['rm', '-rf', self.test_dir])

 RED     = '\033[1;31m'
```

**Comment by ggstone0523** [comment]
Thanks you for comment! I apply change following the comments.

File: tests/runtest.py [hunk 1]
```diff
@@ -801,6 +810,8 @@ def parse_argument():
                         help="Parallel worker count; using all core for default")
     parser.add_argument("-c", "--compiler", dest='compiler', default="gcc",
                         help="Select compiler gcc or clang. (gcc by default)")
+    parser.add_argument("-k", "--keep", dest='keep', action='store_true',
```
Comment (honggyukim) [inline tests/runtest.py:813]: Please change this to `store_false`.

File: tests/runtest.py [hunk 2]
```diff
@@ -801,6 +810,8 @@ def parse_argument():
...
+                        help="Keep compiled binaries")
```
Comment (honggyukim) [inline tests/runtest.py:814]: Please update it as follows.
```
keep the test directories with compiled binaries
```

**Review by honggyukim — COMMENTED** [review]
Sorry, it doesn't work with `os.getcwd()` inside `__init__`. It's better to be like this way.
```diff
diff --git a/tests/runtest.py b/tests/runtest.py
index adb70bd5..07714f4f 100755
--- a/tests/runtest.py
+++ b/tests/runtest.py
@@ -83,7 +83,8 @@ class TestBase:
     TEST_SKIP = -7
     TEST_SUCCESS_FIXED = -8

-    basedir = os.path.dirname(os.getcwd())
+    origdir = os.getcwd()
+    basedir = os.path.dirname(origdir)
     objdir = 'objdir' in os.environ and os.environ['objdir'] or basedir
     uftrace_cmd = objdir + '/uftrace'
     default_opt = '--no-pager --no-event --libmcount-path=' + objdir
@@ -95,7 +96,6 @@ class TestBase:
     def __init__(self, name, result, lang='C', cflags='', ldflags='', sort='task', serial=False):
         _tmp = tempfile.mkdtemp(prefix='test_%s_' % name)
         self.keep = False
-        self.orig_dir = os.getcwd()
         os.chdir(_tmp)
         self.test_dir = _tmp
         self.name = name
@@ -629,7 +629,7 @@ class TestBase:

     def __del__(self):
         if self.keep:
-            sp.call(['mv', self.test_dir, self.orig_dir])
+            sp.call(['mv', self.test_dir, TestBase.origdir])
         else:
             sp.call(['rm', '-rf', self.test_dir])

@@ -807,11 +807,11 @@ def parse_argument():
     parser.add_argument("-t", "--timeout", dest='timeout', default=5,
                         help="fail test if it runs more than TIMEOUT seconds")
     parser.add_argument("-j", "--worker", dest='worker', type=int, default=multiprocessing.cpu_count(),
-                        help="Parallel worker count; using all core for default")
+                        help="parallel worker count; using all core for default")
     parser.add_argument("-c", "--compiler", dest='compiler', default="gcc",
-                        help="Select compiler gcc or clang. (gcc by default)")
-    parser.add_argument("-k", "--keep", dest='keep', action='store_true',
-                        help="Keep compiled binaries")
+                        help="select compiler gcc or clang. (gcc by default)")
+    parser.add_argument("-k", "--keep", dest='keep', action='store_false',
+                        help="keep the test directories with compiled binaries")

     return parser.parse_args()
```

File: tests/runtest.py (hunk 1 above)
Comment (ggstone0523) [reply tests/runtest.py:813]: in my environment, changing 'store_true' to 'store_false' 
cause not keeping test directory when using option -k,  but keeping test directory when not using option -k.

Comment (honggyukim) [reply tests/runtest.py:813]: Ah, right. Please ignore this and keep using `store_true`.

**Comment by ggstone0523** [comment]
Thanks you for detailed review! I apply change following the comment.

**Review by honggyukim — APPROVED** [review]
LGTM. Thanks for your work!

**Pull request merged**